### PR TITLE
[SPARK-31580][BUILD] Upgrade Apache ORC to 1.5.10

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to upgrade Apache ORC to 1.5.10.

### Why are the changes needed?

Apache ORC 1.5.10 is a maintenance release with the following patches.

- [ORC-621](https://issues.apache.org/jira/browse/ORC-621) Need reader fix for ORC-569
- [ORC-616](https://issues.apache.org/jira/browse/ORC-616) In Patched Base encoding, the value of headerThirdByte goes beyond the range of byte
- [ORC-613](https://issues.apache.org/jira/browse/ORC-613) OrcMapredRecordReader mis-reuse struct object when actual children schema differs
- [ORC-610](https://issues.apache.org/jira/browse/ORC-610) Updated Copyright year in the NOTICE file

The following is release note.
- https://issues.apache.org/jira/secure/ReleaseNote.jspa?projectId=12318320&version=12346912

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Pass the Jenkins with the existing ORC tests and a newly added test case.

- The first commit is already tested in `hive-2.3` profile with both native ORC implementation and Hive 2.3 ORC implementation. (https://github.com/apache/spark/pull/28373#issuecomment-620265114)
- The latest run is about to make the test case disable in `hive-1.2` profile which doesn't use Apache ORC.
  - `hive-1.2`: https://github.com/apache/spark/pull/28373#issuecomment-620325906

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala [hunk 1]
```diff
@@ -540,6 +540,12 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
       }
     }
   }
+
+  test("SPARK-31580: Read a file written before ORC-569") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:544]: This test case is in `OrcSuite` and will be tested in both `OrcSourceSuite` and `HiveOrcSourceSuite`.

**Comment by dongjoon-hyun** [comment]
cc @gatorsmile and @yhuai since we need to update ORC dependency in `branch-3.0`.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Hi, @dbtsai .
Could you review this PR?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala [hunk 2]
```diff
@@ -320,4 +320,11 @@ class HiveOrcSourceSuite extends OrcSuite with TestHiveSingleton {
       }
     }
   }
+
+  test("SPARK-31580: Read a file written before ORC-569") {
+    assume(HiveUtils.isHive23) // Hive 1.2 doesn't use Apache ORC
```
Comment (holdenk) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala:325]: I haven't seen this before, that's a cool feature :)
That being said I'm not sure about the version number check here, looking at the related JIRAs https://issues.apache.org/jira/browse/HIVE-22480 & https://issues.apache.org/jira/browse/ORC-569 it seems like it should be fixed earlier? Or is that JIRA wrong?

**Review by holdenk — COMMENTED** [review]
Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala:325]: The Hive JIRA is correct, but all versions in HIVE-22480 are not released officially.
> 1.3.0, 1.2.3, 2.1.2, 2.2.1

From Hive 2.3.x, Hive starts to use Apache ORC. During the transition, Hive side fix seems to be not migrated to Apache ORC correctly.

Comment (holdenk) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcSourceSuite.scala:325]: Thanks for the additional context :) Sounds reasonable.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you for review and approval, @HyukjinKwon .

**Review by holdenk — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you for review and approval, @holdenk !

**Comment by dongjoon-hyun** [comment]
The above one is passed with `-Phive-1.2`.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.0.

**Pull request closed without merging**